File: 백준/Silver/16198. 에너지 모으기/에너지 모으기.py
import sys

input = sys.stdin.readline

n = int(input())
arr = [int(i) for i in input().split()]


def main():
    visited = [False] * len(arr)
    maxv = -float('inf')
    def recur(sumv, cnt):
        nonlocal maxv
        if cnt == len(arr) - 2:
            maxv = max(maxv, sumv)
            return
        
        for i in range(1, len(arr) - 1):
            if not visited[i]:
                left = i - 1
                while visited[left]:
                    left -= 1
                right = i + 1
                while visited[right]:
                    right += 1
                    
                visited[i] = True
                recur(sumv + arr[left] * arr[right], cnt + 1)
                visited[i] = False
    
    recur(0, 0)
    return maxv


print(main())

File: 백준/Silver/16198. 에너지 모으기/에너지 모으기.js
const fs = require('fs');

const read = fs.readFileSync('dev/stdin').toString().trim();

let [n, arr] = read.split("\n");
arr = arr.split(" ").map(Number);

const main = () => {
    const visited = Array.from(arr).fill(false);
    let maxv = 0
    const recur = (sumv, cnt) => {
        if (cnt === arr.length - 2){
            maxv = Math.max(maxv, sumv);
            return;
        }
        
        for (let i = 1; i < arr.length - 1; i++){
            if (visited[i]) continue;
            
            let left = i - 1
            let right = i + 1
            while (visited[left]) left -= 1;
            while(visited[right]) right += 1;
            
            visited[i] = true
            recur(sumv + arr[left] * arr[right], cnt + 1);
            visited[i] = false
        }
    }
    
    recur(0, 0)
    return maxv
}

console.log(main())